accepted/0000-default-command.md

```diff
@@ -0,0 +1,173 @@
+else development of npm is unreasonably limited.
+
+That said, it _does_ impose a non-zero cost in adding new commands to the
+CLI, since users _may_ be relying on a `--default-command` config to handle
+that term in some other way.
+
+### Alternative 1 - run-script _and then_ help
+
+One alternative would be to search for a match among known npm subcommands,
+then `run-script` options, and only _then_ fall back to searching help.
+
+This would be least surprising to current Yarn users, still pretty helpful,
+and not require any opt-in.  However, it is not extensible to other command
+use cases, and still imposes the cost of adding new commands.
+
+### Alternative 2 - just make `help` know about scripts
+
+Another alternative would be to make the current default `help-search`
+behavior look at the `"scripts"` field in the current project's
+`package.json` file.  If the term matches there, it could suggest that the
+user might be thinking of the `run-script` command.
+
+```
+$ npm build
+Top hits for "build"
+————————————————————————————————————————————————————————————————————————————————
+npm help install                                                         build:6
+npm help config                                                          build:5
+npm help explore                                                         build:5
+npm help package-json                                                    build:4
+npm help registry                                                        build:4
+npm help rebuild                                                         build:4
+npm help version                                                         build:4
+npm help developers                                                      build:2
+npm help semver                                                          build:2
+npm help ci                                                              build:1
+————————————————————————————————————————————————————————————————————————————————
+(run with -l or --long to see more context)
+
+`npm run build` will execute the following command:
+  node scripts/build-my-stuff.js
+```
+
+This provides some guidance to users who expect `run-script` to be the
+default command, without requiring much magic, and without the added cost
+of adding new commands in the future.
+
+### Why this, then?
+
+Making the `default-command` configurable also enables a few other
+interesting use cases:
+
+- `npm config set default-command exec` This turns npm into something like
+  a more powerful `npx`.  You can do `npm install` to install your
+  dependencies, `npm create-react-app` to init a new react app, or `npm
+  rimraf folder` to remove a folder recursively.
+- `npm config set default-command install` Since `npm install` is the most
+  common use case for npm, why not just drop the `install`?  Run `npm` with
+  no args to install all your dependencies, or `npm react --save-peer` to
+  install react and save as a peer dependency.
```

Comment:
this seems like it would be really confusing for a non-advanced user who uses a project with default-command set in `.npmrc`.

Maybe it could provide some kind of output like:
```sh
> npm foo
Running `npm run-script foo`...
"Output of npm foo here"

> npm foo --silent
"Output of npm foo here"
```

Reply:
That's a good idea.  Could also probably want to have a `npm.log.notice('default command notice blah')` so that it automatically gets placed in the debug log when stuff goes sideways.